Extend AddMethod

ssiccha opened the pull request:
Change AddMethod to now take two arguments: the method database and a record which contains the method to be added and information about it. 

Setting the InfoClass InfoAddMethod to 1 now produces messages when AddMethod is called with unbound method.validatesInput component.

I added a few small changes to the documentation. The PR is now ready for review.

Now we can address #107 method for method.

wilfwilson commented:
How is the `method.validatesInput` going to be useful? How are you going to use it? (Is it going to be documented, etc?)

wilfwilson reviewed (commented):
I'm fine with `AddMethod` taking a record rather than various arguments. I'm just not sure what you intend for the `validatesInput` component, which I asked about above.

ssiccha commented:
> How is the `method.validatesInput` going to be useful? How are you going to use it? (Is it going to be documented, etc?)

I'm going to use it to work on #107. E.g. for each homomorphism I'm going to make it explicit, whether it checks its input or not. For now I'd rather see this as a sort of meta-information on the homomorphism. We could later require in `AddMethod` the method to have a `method.validatesInput`, to make sure that the creator of the method did not simply forget about the input validation, as it apparently has happened with some methods already.

ssiccha commented:
If we were to go with requiring it in `AddMethod` we could document that in the "what is a method" section.

ssiccha commented:
> For now I'd rather see this as a sort of meta-information on the homomorphism. 

I'll put this information into the PR.

wilfwilson commented:
Good plan.

ssiccha commented:
Thanks for the review! :)

ssiccha commented:
Small update: I renamed `validatesInput` to `validatesOrAlwaysValidInput`. I think there are some "homomorphisms" which do not need to check their input. For example the plain StabChain method shouldn't have to. I'll double-check though.

Another small change. I added some commas in the commit `base: AddMethod input now is database and a record`.

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -20,28 +20,28 @@
+    if not IsBound(method.comment) then
+        method.comment := "";
+    fi;
+    # method.validatesOrAlwaysValidInput for now only stores meta-information.
```
I do find `validatesOrAlwaysValidInput` rather clunky, but since it's not yet used, I'll not object to it here; it'd be trivial to rename later on anyway.
